### PR TITLE
CPos.ToMPos, get and calculate x and y once.

### PR DESCRIPTION
**Pull request by anvilvapre**
Old IL:
```
  .method public hidebysig instance default OpenRA.MPos ToMPos(OpenRA.MapGridType gridType) cil managed
  {                                                                             
    // Method begins at Relative Virtual Address (RVA) 0x35D8                   
    // Code size 59 (0x3B)                                                      
    .maxstack 2                                                                 
    .locals init(int32 V_0, int32 V_1)                                          
    IL_0000: ldarg.1                                                            
    IL_0001: brtrue.s     IL_0015                                               
    IL_0003: ldarg.0                                                            
    IL_0004: call instance int32 class OpenRA.CPos::get_X()                     
    IL_0009: ldarg.0                                                            
    IL_000a: call instance int32 class OpenRA.CPos::get_Y()                     
    IL_000f: newobj instance void class OpenRA.MPos::.ctor(int32, int32)        
    IL_0014: ret                                                                
    IL_0015: ldarg.0                                                            
    IL_0016: call instance int32 class OpenRA.CPos::get_X()                     
    IL_001b: ldarg.0                                                            
    IL_001c: call instance int32 class OpenRA.CPos::get_Y()                     
    IL_0021: sub                                                                
    IL_0022: ldc.i4.2                                                           
    IL_0023: div                                                                
    IL_0024: stloc.0                                                            
    IL_0025: ldarg.0                                                            
    IL_0026: call instance int32 class OpenRA.CPos::get_X()                     
    IL_002b: ldarg.0                                                            
    IL_002c: call instance int32 class OpenRA.CPos::get_Y()                     
    IL_0031: add                                                                
    IL_0032: stloc.1                                                            
    IL_0033: ldloc.0                                                            
    IL_0034: ldloc.1                                                            
    IL_0035: newobj instance void class OpenRA.MPos::.ctor(int32, int32)        
    IL_003a: ret                                                                
  } // End of method OpenRA.MPos OpenRA.CPos::ToMPos(OpenRA.MapGridType) 
```

### CODE REVIEW

**Comment by pchote** [comment]
This gets called a lot so I was expecting to see some kind of change in the tick time from this, but comparing the histogram of tick_time over the first 5000 ticks of the RA shellmap showed that any potential improvement here is smaller than the variation between my three test runs before/after.

I guess the runtime JIT is optimizing these extra calls away?

**Comment by anvilvapre** [comment]
Doesn't seem to be the case on my machine, Linux. I don't mind to close this pull request.

Bleed:

`get_X`

```
00007f2efa69de50 55                   push    rbp
00007f2efa69de51 4883ec10             sub     rsp, 0x10
00007f2efa69de55 488d6c2410           lea     rbp, [rsp + 0x10]
00007f2efa69de5a 48897df8             mov     qword ptr [rbp - 0x8], rdi
00007f2efa69de5e 48b8c0bd34f82e7f0000 movabs  rax, 0x7f2ef834bdc0
00007f2efa69de68 833800               cmp     dword ptr [rax], 0x0
00007f2efa69de6b 7405                 je      0x7f2efa69de72
00007f2efa69de6d e8fe76ef76           call    0x7f2f71595570 (JitHelp: CORINFO_HELP_DBG_IS_JUST_MY_CODE)
00007f2efa69de72 488b45f8             mov     rax, qword ptr [rbp - 0x8]
00007f2efa69de76 8b00                 mov     eax, dword ptr [rax]
00007f2efa69de78 c1f814               sar     eax, 0x14
00007f2efa69de7b 488d6500             lea     rsp, [rbp]
00007f2efa69de7f 5d                   pop     rbp
00007f2efa69de80 c3                   ret     
```
`get_Y`
```
00007f2efa69dea0 55                   push    rbp
00007f2efa69dea1 4883ec10             sub     rsp, 0x10
00007f2efa69dea5 488d6c2410           lea     rbp, [rsp + 0x10]
00007f2efa69deaa 48897df8             mov     qword ptr [rbp - 0x8], rdi
00007f2efa69deae 48b8c0bd34f82e7f0000 movabs  rax, 0x7f2ef834bdc0
00007f2efa69deb8 833800               cmp     dword ptr [rax], 0x0
00007f2efa69debb 7405                 je      0x7f2efa69dec2
00007f2efa69debd e8ae76ef76           call    0x7f2f71595570 (JitHelp: CORINFO_HELP_DBG_IS_JUST_MY_CODE)
00007f2efa69dec2 488b45f8             mov     rax, qword ptr [rbp - 0x8]
00007f2efa69dec6 8b00                 mov     eax, dword ptr [rax]
00007f2efa69dec8 c1f804               sar     eax, 0x4
00007f2efa69decb 480fbfc0             movsx   rax, ax
00007f2efa69decf c1f804               sar     eax, 0x4
00007f2efa69ded2 488d6500             lea     rsp, [rbp]
00007f2efa69ded6 5d                   pop     rbp
00007f2efa69ded7 c3                   ret  
```
`ToMPos`

```
00007f2efa69dd30 55                   push    rbp                                                                                                                                             
00007f2efa69dd31 4883ec50             sub     rsp, 0x50                                                                                                                                       
00007f2efa69dd35 488d6c2450           lea     rbp, [rsp + 0x50]                                                                                                                               
00007f2efa69dd3a c4413857c0           vxorps  xmm8, xmm8, xmm8                                                                                                                                
00007f2efa69dd3f c5797f45e0           vmovdqa xmmword ptr [rbp - 0x20], xmm8                                                                                                                  
00007f2efa69dd44 33c0                 xor     eax, eax                                                                                                                                        
00007f2efa69dd46 8945f0               mov     dword ptr [rbp - 0x10], eax                                                                                                                     
00007f2efa69dd49 48897df8             mov     qword ptr [rbp - 0x8], rdi                                                                                                                      
00007f2efa69dd4d 8975f4               mov     dword ptr [rbp - 0xc], esi                                                                                                                      
00007f2efa69dd50 48b8c0bd34f82e7f0000 movabs  rax, 0x7f2ef834bdc0                                                                                                                             
00007f2efa69dd5a 833800               cmp     dword ptr [rax], 0x0                                                                                                                            
00007f2efa69dd5d 7405                 je      0x7f2efa69dd64                                                                                                                                  
00007f2efa69dd5f e80c78ef76           call    0x7f2f71595570 (JitHelp: CORINFO_HELP_DBG_IS_JUST_MY_CODE)                                                                                      
00007f2efa69dd64 90                   nop                                                                                                                                                     

/OpenRA/OpenRA.Game/CPos.cs @ 70:                                                                                                                                         
00007f2efa69dd65 837df400             cmp     dword ptr [rbp - 0xc], 0x0                                                                                                                      
00007f2efa69dd69 400f94c7             sete    dil                                                                                                                                             
00007f2efa69dd6d 400fb6ff             movzx   edi, dil                                                                                                                                        
00007f2efa69dd71 897de8               mov     dword ptr [rbp - 0x18], edi                                                                                                                     
00007f2efa69dd74 837de800             cmp     dword ptr [rbp - 0x18], 0x0                                                                                                                     
00007f2efa69dd78 7438                 je      0x7f2efa69ddb2                                                                                                                                  

/OpenRA/OpenRA.Game/CPos.cs @ 71:                                                                                                                                         
00007f2efa69dd7a 488b7df8             mov     rdi, qword ptr [rbp - 0x8]                                                                                                                      
00007f2efa69dd7e e80dd629fe           call    0x7f2ef893b390 (OpenRA.CPos.get_X(), mdToken: 000000000600005B)                                                                                 
00007f2efa69dd83 8945c4               mov     dword ptr [rbp - 0x3c], eax                                                                                                                     
00007f2efa69dd86 488b7df8             mov     rdi, qword ptr [rbp - 0x8]                                                                                                                      
00007f2efa69dd8a e809d629fe           call    0x7f2ef893b398 (OpenRA.CPos.get_Y(), mdToken: 000000000600005C)                                                                                 
00007f2efa69dd8f 8945c0               mov     dword ptr [rbp - 0x40], eax                                                                                                                     
00007f2efa69dd92 33ff                 xor     edi, edi                                                                                                                                        
00007f2efa69dd94 48897db8             mov     qword ptr [rbp - 0x48], rdi                                                                                                                     
00007f2efa69dd98 488d7db8             lea     rdi, [rbp - 0x48]                                                                                                                               
00007f2efa69dd9c 8b75c4               mov     esi, dword ptr [rbp - 0x3c]                                                                                                                     
00007f2efa69dd9f 8b55c0               mov     edx, dword ptr [rbp - 0x40]                                                                                                                     
00007f2efa69dda2 e8d9e1b4fe           call    0x7f2ef91ebf80 (OpenRA.MPos..ctor(Int32, Int32), mdToken: 00000000060003A9)                                                                     
00007f2efa69dda7 488b7db8             mov     rdi, qword ptr [rbp - 0x48]                                                                                                                     
00007f2efa69ddab 48897de0             mov     qword ptr [rbp - 0x20], rdi                                                                                                                     
00007f2efa69ddaf 90                   nop                                                                                                                                                     
00007f2efa69ddb0 eb70                 jmp     0x7f2efa69de22                                                                                                                                  

/OpenRA/OpenRA.Game/CPos.cs @ 83:                                                                                                                                         
00007f2efa69ddb2 488b7df8             mov     rdi, qword ptr [rbp - 0x8]                                                                                                                      
00007f2efa69ddb6 e8d5d529fe           call    0x7f2ef893b390 (OpenRA.CPos.get_X(), mdToken: 000000000600005B)                                                                                 
00007f2efa69ddbb 8945dc               mov     dword ptr [rbp - 0x24], eax                                                                                                                     
00007f2efa69ddbe 488b7df8             mov     rdi, qword ptr [rbp - 0x8]                                                                                                                      
00007f2efa69ddc2 e8d1d529fe           call    0x7f2ef893b398 (OpenRA.CPos.get_Y(), mdToken: 000000000600005C)                                                                                 
00007f2efa69ddc7 8945d8               mov     dword ptr [rbp - 0x28], eax                                                                                                                     
00007f2efa69ddca 8b7ddc               mov     edi, dword ptr [rbp - 0x24]                                                                                                                     
00007f2efa69ddcd 2b7dd8               sub     edi, dword ptr [rbp - 0x28]                                                                                                                     
00007f2efa69ddd0 897db4               mov     dword ptr [rbp - 0x4c], edi                                                                                                                     
00007f2efa69ddd3 8b7db4               mov     edi, dword ptr [rbp - 0x4c]                                                                                                                     
00007f2efa69ddd6 c1ef1f               shr     edi, 0x1f                                                                                                                                       
00007f2efa69ddd9 037db4               add     edi, dword ptr [rbp - 0x4c]                                                                                                                     
00007f2efa69dddc d1ff                 sar     edi                                                                                                                                             
00007f2efa69ddde 897df0               mov     dword ptr [rbp - 0x10], edi                                                                                                                     

/OpenRA/OpenRA.Game/CPos.cs @ 84:                                                                                                                                         
00007f2efa69dde1 488b7df8             mov     rdi, qword ptr [rbp - 0x8]                                                                                                                      
00007f2efa69dde5 e8a6d529fe           call    0x7f2ef893b390 (OpenRA.CPos.get_X(), mdToken: 000000000600005B)                                                                                 
00007f2efa69ddea 8945d4               mov     dword ptr [rbp - 0x2c], eax                                                                                                                     
00007f2efa69dded 488b7df8             mov     rdi, qword ptr [rbp - 0x8]                                                                                                                      
00007f2efa69ddf1 e8a2d529fe           call    0x7f2ef893b398 (OpenRA.CPos.get_Y(), mdToken: 000000000600005C)                                                                                 
00007f2efa69ddf6 8945d0               mov     dword ptr [rbp - 0x30], eax                                                                                                                     
00007f2efa69ddf9 8b7dd4               mov     edi, dword ptr [rbp - 0x2c]                                                                                                                     
00007f2efa69ddfc 037dd0               add     edi, dword ptr [rbp - 0x30]                                                                                                                     
00007f2efa69ddff 897dec               mov     dword ptr [rbp - 0x14], edi                                                                                                                     

/OpenRA/OpenRA.Game/CPos.cs @ 85:                                                                                                                                         
00007f2efa69de02 33ff                 xor     edi, edi                                                                                                                                        
00007f2efa69de04 48897dc8             mov     qword ptr [rbp - 0x38], rdi                                                                                                                     
00007f2efa69de08 488d7dc8             lea     rdi, [rbp - 0x38]                                                                                                                               
00007f2efa69de0c 8b75f0               mov     esi, dword ptr [rbp - 0x10]                                                                                                                     
00007f2efa69de0f 8b55ec               mov     edx, dword ptr [rbp - 0x14]                                                                                                                     
00007f2efa69de12 e869e1b4fe           call    0x7f2ef91ebf80 (OpenRA.MPos..ctor(Int32, Int32), mdToken: 00000000060003A9)                                                                     
00007f2efa69de17 488b45c8             mov     rax, qword ptr [rbp - 0x38]                                                                                                                     
00007f2efa69de1b 488945e0             mov     qword ptr [rbp - 0x20], rax                                                                                                                     
00007f2efa69de1f 90                   nop                                                                                                                                                     
00007f2efa69de20 eb00                 jmp     0x7f2efa69de22                                                                                                                                  

/OpenRA/OpenRA.Game/CPos.cs @ 86:                                                                                                                                         
00007f2efa69de22 488b45e0             mov     rax, qword ptr [rbp - 0x20]                                                                                                                     
00007f2efa69de26 488d6500             lea     rsp, [rbp]                                                                                                                                      
00007f2efa69de2a 5d                   pop     rbp                                                                                                                                             
00007f2efa69de2b c3                   ret                                           

```

In general I think it would also be a good idea to set PerfTimer (optionally PerfSampler) by default to microseconds, or at least more fine grained than milliseconds. Only high up the code things typically still take milliseconds. It may also give more insight in the stability and therefore quality of runs.

**Comment by teinarss** [comment]
For these kinds of performance testing I would really recommend
benchmarkdotnet.

**Comment by teinarss** [comment]
Took the effort to try it in [Sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACATgUzABMB7AOwBsBPbCYXAVwGNhsBhABRImxGwEkAogEcGYYGABGFfAB5O3AHzoAkAG9VKgPRb2JEriIBLMuPw8wBbAAcwTANb4i2E9jDZMCbJKOsIRgHMyJxcyYE0dbAANNzJnAE03KyQvH2AeAAp/IKcASj4AWgQABgAWAA4AOmqS0oB2CN0AGTAqfFwXCzJsBjJs4OdJKmB8RuwOO3sTAM7oqPm5hfmF+NXEtfW17Cadpu3dzSw8QlJKGhNWACFfCABudDGYox5bIiIQ4BJsAFkAZUuADS0L5pWgACyMADNWKRzNhgGD8NgmAYCCxaIFuvgAB4jPpGciHHAXaLYAC8imw13S2EUlJK9zQYwAghRMdhEgB3XxgtzggzAIFMSCC+GI7oQCHQsVIgh0bAkSEy7Cc1pjAAqiIxOWckoF3l84KhMJIcIRSJRuDRfnZOLx/kJaBUKiOJMSFOwGSyYIFuQy1J4dOwpVyuVplNKjKJ3mGSJabQ6HoyQxGuQDUadLpwCggGRJaQgYbUVJu5IN6Vu2AAvtGc3mwthsUCSVRcpoVHwEc8Mk3sFQgcUi9Xa1xcyTey2gSmkRRWu025mNJmVAGyz3sAAybDFbEAMX3YdksmwJWwME9NC3O/3u8Px/KZ+ws4TjOdNaZH+jcCQADZvPoKAVax2nEAxyTJDIc2wKB8CBKCSHNXAhzgOpoPwSpVzJMkFUQjCbkrd9nSOb8/0kACgJAz4OgAQggqCYLg0ccMRJDsGLFC0LwmlaOY9ouLuYcPyInASAAN3aXAjHeUJWAAcXwYAAAlIDBNgSHeDJkNQgNKnkpSVLUjTcgIh5MyOMiSEAkQxAoXN4MQrSSxpLDeNwfiTLMkTxKtKSkQsqzRDAWyMhISQACt8HRUKwsc6LZigjct2soLc0g0dcmi4zBOjMSJN87Bv2KbB1RIX56GmTS2IK1CYgAamwAAiAEGuwer4g86NviYkquu4DJvjAaxZMkoh1SoYDsACEaxuAhdnSXZ0VChT0pqkmakRcgahum8b0IAJUiiQyACBhZyQ9sXVQ4JOR+UcMiiIF4mM0zFsiNSyG81hIVwEgoGwA6WDAY7TssfgIF+hTJKYZF8AoQCeyBVsbG4XwCW6T5jkB4Gzugwbkf8YA0c9BggVEubnUibACmKrU6DAAIAnaEIfs5HgoDARwxWebBISkohqE9CEwj4IhcDAG73GMemxagajye0XQMjAMN3v8d4OnNY5Z0J8hJSMawIBAdtKep2qkGxT18HE7oCkpdSiDDYgiEyYogSQMMMYyEnsDJ43dCp1rzc9e2qcpK38DIR23kyJA3Y9r4vdJ+WTcDmgMnDm27beKPnc9ApY+wd34QT73feXFOzbTkPbewDOc5duPi+JpO/c9SQw01JnIQMfAjfL/3qbAC36qGNxo89dxqfbrQvAxhhg7OWvrdCJgCBgsIeBRXpgGTge3GHmMx9z9wR6b0T21EyxGzLKJX2dS+OhobD2ovq/5+w9dqaR3QEDvlQH59mWA+VA/4cWurdPqpdnqZnfNgOB8C4HoAQQgo4vUeA9VHAgfqg1hprV2pNHas1NALWdMtDIq1Rr4M2jgwh+1DpAxOmdeWl1sDgLQfdR60DW7vU+jzH6f0AZHUYaDcGMFyrQyYLDeGvYkbWBRjrdGXxrQMJBh0dm1h8ao3IM3H2u84HU07rQCQDMmbOBZmzDmSIuw8F5m8AWGQhagGwKLcWfIpZTTALLPRSsVa618hrLUBBtZoz1gbPur095mwtunZeNd7b109K7Qu8cdFlwifowO0Tq5h2tgkjIBci6eyga3AOldLaxKzg7I+mR86NyKS3fuGSykhTeKHJeEc8lJMKSXBp6TkwdxYvgbuBBwkUz3kPVqh8naZEnt4XIM8m7zxCovDOK814RxpFvMIeiA4TJHjQaZfJT4Y3PmgZB5yfZvzXDEamT1sA/z/gA0SN9Jkv2XGA/AN12FQLvu+VQsD4FHAjgwP6W1cGUImsWQRKizpAmhdjERENxHZSdEcAgxByACzoIwdEaDeACGShIaQcg0HKCdCQtFJxMXnAbAAVSBAANQzMJCBY4GzexJGTKqtKywMErAyssokOpCSzAVX8/5LIUTFlRcC/UmIMVZa5RyMFKg8pcghFiqrNxbhVQK9VuEmWCRZSRCVgESDAWlWBWicruBoSBHijV84qocWohkGC4ElXCpZblHy0kSR6WUpKQy+BKrsVQrS3SClA2qXUiGsMAA9bADLI36SDbGzSXrRX+WwMlYKDqHLOtQo6xM2FrEeW9d5EaflyK5tzNFQ6CpwqxXCrMPFiUc2BWCjawsmVM1HB9VWsVRUSplUksdUN1VsA8vqk1Fq9VDV/M/J5dg3USB1jBbQgheCiGLnbGQih60PUbu3XQrGwjzrLhYeAus9Kk1cMaTw9oX1+GYyEao3GGi5EEyJhjeF56wZIqhjDOGmiFElIMbTYxjMCBmJIKzXGnNrE8z5vYxxIsxYS2cUYaWni5atx8XofE6tlRBPECEiEYTwOByWasuJ2dqmegLvnFJCM+w7NNkgZ5MSI5tPiQx/JA4WMyPY4HLj2T2mR3410oTiNvHtxpl3HuoyFYZLAPPeqCBJDPMOUrSZ8nPbCao+4S01oiaQjsDK7mTsQjLWedze2r8OiKkhBABSa4BVbiLi5JA2AAD8hd8XFEeVfJ+noBXU2c65ne9yTwB1pcFjoFtsLhb7KAq6nyV19WE78/56AgA==) and as far as i can tell it looks like the JIT is smart enough to optimize this, compare ToMPos to ToMPos2.

File: OpenRA.Game/CPos.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public CPos(int x, int y, byte layer)
 		public static CVec operator -(CPos a, CPos b) { return new CVec(a.X - b.X, a.Y - b.Y); }
 
 		public static bool operator ==(CPos me, CPos other) { return me.Bits == other.Bits; }
-		public static bool operator !=(CPos me, CPos other) { return !(me == other); }
+		public static bool operator !=(CPos me, CPos other) { return me.Bits != other.Bits; }
```
Comment (abcdefg30) [inline OpenRA.Game/CPos.cs:54]: This change also didn't show a difference on the sharplab page @teinarss suggested. Both versions resulted in
```
CPos.op_Inequality(CPos, CPos)
    L0000: cmp ecx, edx
    L0002: setne al
    L0005: movzx eax, al
    L0008: ret
```

**Comment by abcdefg30** [comment]
> I don't mind to close this pull request.

Seems like there is indeed nothing to gain here?

**Comment by anvilvapre** [comment]
Curious though to why the calls are there in my earlier assembly listing. Possibly because make builds in debug mode? Are optimization flags only set for some build targets?

**Comment by pchote** [comment]
IL is just the first level of compilation. The runtime JITs this to real native assembly, with further optimizations able to take place.

**Comment by anvilvapre** [comment]
I guess it is because of debug mode. If in sharplib you select debug in the pull down menu on the right side you will see the function calls. Code similar to my earlier paste.

**Comment by anvilvapre** [comment]
Yes, but as sharplib shows JIT-ASM in debug mode is not optimized. 

So in final releases you may want to run in debug mode - to have symbols and allow for easy bug fixes - but then you still would like dotnet to try to optimize the code. In C++ the difference between a -g and -O flag.

**Comment by teinarss** [comment]
> 
> 
> Curious though to why the calls are there in my earlier assembly listing. Possibly because make builds in debug mode? Are optimization flags only set for some build targets?

By default the Optimize code is turned off in debug build, thats why tools like benchmarkdotnet will give you a warning when running in debug build.

**Pull request closed without merging**